virtio-queue: improve testing

lauralt opened the pull request:
This PR adds or improves tests for validating the notification suppression mechanism, the reset, indirect descriptor requirements and processing of them.

andreeaflorescu commented on crates/virtio-queue/src/lib.rs:
```diff
@@ -1182,12 +1182,38 @@ mod tests {
+        let m = &GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x100000)]).unwrap();
+        let vq = MockSplitQueue::new(m, 16);
+
+        for j in 0..15 {
```
nit: Why use j?

lauralt replied:
Just some copy paste from another test, I changed it with `i`.

andreeaflorescu commented on crates/virtio-queue/src/lib.rs:
```diff
@@ -1182,12 +1182,38 @@ mod tests {
     }
 
+    #[test]
+    fn test_ttl_wrap_around() {
```
I am not sure why is this called `wrap_around`. Can you add some more details about the test case that you're adding here?

lauralt replied:
Added some comments in the test.

andreeaflorescu commented on crates/virtio-queue/src/lib.rs:
```diff
@@ -1493,11 +1567,35 @@ mod tests {
+        assert_eq!(q.state.size, 8);
+        // `create_queue` also marks the queue as ready.
+        assert_eq!(q.state.ready, true);
+        assert_ne!(q.state.desc_table, GuestAddress(DEFAULT_DESC_TABLE_ADDR));
+        assert_ne!(q.state.avail_ring, GuestAddress(DEFAULT_AVAIL_RING_ADDR));
+        assert_ne!(q.state.used_ring, GuestAddress(DEFAULT_USED_RING_ADDR));
+        assert_ne!(q.state.next_avail, Wrapping(0));
+        assert_ne!(q.state.next_used, Wrapping(0));
+        assert_ne!(q.state.signalled_used, None);
```
Why do we check these before and after the reset? They seem to have the same values.

lauralt replied:
I am using `assert_ne!` before the reset.

andreeaflorescu commented on crates/virtio-queue/src/lib.rs:
```diff
@@ -1544,6 +1642,11 @@ mod tests {
         assert_eq!(q.needs_notification().unwrap(), true);
         assert_eq!(q.needs_notification().unwrap(), false);
+
+        m.write_obj::<u16>(u16::MAX - 3, avail_addr.unchecked_add(4 + qsize as u64 * 2))
```
Can you add a comment here to explain why you're calling write again & needs notification is true?

lauralt replied:
done

andreeaflorescu commented on crates/virtio-queue/src/lib.rs:
```diff
@@ -1582,4 +1685,154 @@ mod tests {
+
+            while let Some(_chain) = q.iter().unwrap().next() {
+                // Do something with the chain. Here the device would consume entries from the
+                // available ring, add an entry in the used ring and optionally notify the driver.
```
I would also add a comment that for the purpose of this test what happens with the chain is not relevant, so we just ignore it as we just want to consume the entries in the queue.

lauralt replied:
done

andreeaflorescu reviewed (commented):
Can you add some more comments in the tests? For some of them it's not obvious why we're adding them, so maybe we can give a short description at the beginning (i.e. this is a negative test, its purpose it to check that in case of  misconfiguration (name the misconfiguration) the device does not cause a panic).

Can you also update the commit message? I think you were saying that some of the tests that we already have were not following the specification. We should name them in the commit message to have a history. We should also say in the commit message what kind of tests we're adding and why. What is particularly interesting about the tests is that we are adding negative tests checking what happens when the driver is not following the spec (i.e. the things with MUST).

Just saying in the commit message that we're "improving testing" is an understatement ;)

lauralt commented:
> Can you add some more comments in the tests? For some of them it's not obvious why we're adding them, so maybe we can give a short description at the beginning (i.e. this is a negative test, its purpose it to check that in case of misconfiguration (name the misconfiguration) the device does not cause a panic).

I added more comments.

> 
> Can you also update the commit message? I think you were saying that some of the tests that we already have were not following the specification. We should name them in the commit message to have a history. We should also say in the commit message what kind of tests we're adding and why. What is particularly interesting about the tests is that we are adding negative tests checking what happens when the driver is not following the spec (i.e. the things with MUST).

I updated the commit message. I'm not fixing yet the tests that were not following the spec. This was part of [this issue](https://github.com/rust-vmm/vm-virtio/issues/95) and I'll handle it in a separate PR.

> 
> Just saying in the commit message that we're "improving testing" is an understatement ;)

lauralt commented:
> > Can you add some more comments in the tests? For some of them it's not obvious why we're adding them, so maybe we can give a short description at the beginning (i.e. this is a negative test, its purpose it to check that in case of misconfiguration (name the misconfiguration) the device does not cause a panic).
> 
> I added more comments.
> 
> > Can you also update the commit message? I think you were saying that some of the tests that we already have were not following the specification. We should name them in the commit message to have a history. We should also say in the commit message what kind of tests we're adding and why. What is particularly interesting about the tests is that we are adding negative tests checking what happens when the driver is not following the spec (i.e. the things with MUST).
> 
> I updated the commit message. I'm not fixing yet the tests that were not following the spec. This was part of [this issue](https://github.com/rust-vmm/vm-virtio/issues/95) and I'll handle it in a separate PR.
> 
> > Just saying in the commit message that we're "improving testing" is an understatement ;)

I added a separate commit for https://github.com/rust-vmm/vm-virtio/issues/95 and also did some fixes to the already existing test for indirect descriptors. More details in the commit message/comments.

slp reviewed (approved):
LGTM, thanks!
